error in cgen with struct

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: linux, Pop!_OS 21.10
Processor: 8 cpus, 64bit, little endian, 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
CC version: cc (Ubuntu 11.2.0-7ubuntu2) 11.2.0

getwd: /home/will/Code/RCBot
vmodules: /home/will/.vmodules
vroot: /home/will/.local/v
vexe: /home/will/.local/v/v
vexe mtime: 2022-03-19 20:40:05
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 3f35103.c05634e

Git version: git version 2.32.0
Git vroot status: weekly.2022.11-32-gc05634eb
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 3654d6a8

```
full project: https://codeberg.org/wbehrens/RCBot/src/branch/vi2c

**What did you do?**
`v -cg -cc gcc -path 'modules|@vmodules|@vlib|' -o rcbot modules/main`
main.v:
```v
module main

import pca9685
import time

const(
	name = 'WaveShare DC Motor Controller'
	address = byte(0x40)
	bus_path = '/dev/i2c-1'
)

/*
self.PWMA = 0
self.AIN1 = 1
self.AIN2 = 2
self.PWMB = 5
self.BIN1 = 3
self.BIN2 = 4
*/

struct Motor {
	pwm int
	in_1 int
	in_2 int
mut:
	cntl pca9685.Device
}

fn (mut self Motor) run(speed int, invert bool) {
	if speed > 100 {
		self.cntl.set_dutycycle(self.pwm, 100) or { panic(err) }
	} else {
		self.cntl.set_dutycycle(self.pwm, speed) or { panic(err) }
	}

	if !invert { // Forward
		self.cntl.set_level(self.in_1, 0) or { panic(err) }
		self.cntl.set_level(self.in_2, 1) or { panic(err) }
	} else { // Backward
		self.cntl.set_level(self.in_1, 1) or { panic(err) }
		self.cntl.set_level(self.in_2, 0) or { panic(err) }
	}
}

fn (mut self Motor) stop() {
    self.cntl.set_dutycycle(self.pwm, 0) or { panic(err) }
}

fn main() {
	println('Motor Test')
	mut motor := Motor{
		pwm: 0
		in_1: 1
		in_2: 2
		cntl: pca9685.new_device('cntl', 0x40, '/dev/i2c-1')?
	}

	println('Forward 2 sec')
	motor.run(100, false)
	time.sleep(2)

	println('Backward 2 sec')
	motor.run(100, true)
	time.sleep(2)

	println('Stop')
	motor.stop()
}


```

**What did you expect to see?**

The code to control a brushless dc motor over i2c


**What did you see instead?**
```
/tmp/v_1000/rcbot.9321354038293395484.tmp.c:1442:25: error: field ‘cntl’ has incomplete type
 1442 |         pca9685__Device cntl;
      |                         ^~~~
/tmp/v_1000/rcbot.9321354038293395484.tmp.c: In function ‘vire__i2c__new_device’:
/tmp/v_1000/rcbot.9321354038293395484.tmp.c:17731:217: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
17731 |         if (vire__wrapper__ioctl(self.bus.fd, (*(byte*)map_get(ADDR(map, _const_vire__i2c__ioctl_cmds), &(string[]){_SLIT("slave")}, &(byte[]){ 0 })), new_array_from_c_array(1, 1, sizeof(voidptr), _MOV((voidptr[1]){((voidptr)(self.address))}))) < 0) {
      |                                                                                                                                                                                                                         ^
/tmp/v_1000/rcbot.9321354038293395484.tmp.c: In function ‘vire__i2c__Device_read_reg_byte’:
/tmp/v_1000/rcbot.9321354038293395484.tmp.c:17770:70: error: incompatible type for argument 1 of ‘array_get_with_check’
17770 |         byte* _t3 = (byte*)/*ee elem_ptr_typ */(array_get_with_check(vire__i2c__Device_read_reg(self, reg, 1), 0));
      |                                                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                                                      |
      |                                                                      Option_Array_byte
/tmp/v_1000/rcbot.9321354038293395484.tmp.c:7304:36: note: expected ‘array’ but argument is of type ‘Option_Array_byte’
 7304 | voidptr array_get_with_check(array a, int i) {
      |                              ~~~~~~^
builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -4524,17 +4524,17 @@ fn (mut g Gen) write_builtin_types() {
 // Sort the types, make sure types that are referenced by other types
 // are added before them.
 fn (mut g Gen) write_sorted_types() {
+	g.type_definitions.writeln('// #start sorted_symbols')
+	defer {
+		g.type_definitions.writeln('// #end sorted_symbols')
+	}
 	mut symbols := []&ast.TypeSymbol{cap: g.table.type_symbols.len} // structs that need to be sorted
 	for sym in g.table.type_symbols {
 		if sym.name !in c.builtins {
 			symbols << sym
 		}
 	}
-	// sort structs
 	sorted_symbols := g.sort_structs(symbols)
-	// Generate C code
-	g.type_definitions.writeln('// builtin types:')
-	g.type_definitions.writeln('//------------------ #endbuiltin')
 	g.write_types(sorted_symbols)
 }
 
@@ -4706,7 +4706,7 @@ fn (mut g Gen) write_types(symbols []&ast.TypeSymbol) {
 }
 
 // sort structs by dependant fields
-fn (g &Gen) sort_structs(typesa []&ast.TypeSymbol) []&ast.TypeSymbol {
+fn (mut g Gen) sort_structs(typesa []&ast.TypeSymbol) []&ast.TypeSymbol {
 	util.timing_start(@METHOD)
 	defer {
 		util.timing_measure(@METHOD)
@@ -4742,12 +4742,23 @@ fn (g &Gen) sort_structs(typesa []&ast.TypeSymbol) []&ast.TypeSymbol {
 					field_deps << dep
 				}
 				for field in sym.info.fields {
-					dep := g.table.sym(field.typ).name
+					if field.typ.is_ptr() {
+						continue
+					}
+					fsym := g.table.sym(field.typ)
+					dep := fsym.name
 					// skip if not in types list or already in deps
-					if dep !in type_names || dep in field_deps || field.typ.is_ptr() {
+					if dep !in type_names || dep in field_deps {
 						continue
 					}
 					field_deps << dep
+					if fsym.info is ast.Alias {
+						xdep := g.table.sym(fsym.info.parent_type).name
+						if xdep !in type_names || xdep in field_deps {
+							continue
+						}
+						field_deps << xdep
+					}
 				}
 			}
 			// ast.Interface {}

--- a/vlib/v/tests/modules/alias_to_another_module/alias.v
+++ b/vlib/v/tests/modules/alias_to_another_module/alias.v
@@ -1,0 +1,9 @@
+module alias_to_another_module
+
+import another_module
+
+pub type MyAlias = another_module.SomeStruct
+
+pub fn (m MyAlias) alias_method() int {
+	return 42
+}

--- a/vlib/v/tests/modules/another_module/module.v
+++ b/vlib/v/tests/modules/another_module/module.v
@@ -1,0 +1,12 @@
+module another_module
+
+pub struct SomeStruct {
+pub mut:
+	x int
+	y int
+	z int
+}
+
+pub fn (s SomeStruct) some_method() int {
+	return 999 + s.x + s.y + s.z
+}

--- a/vlib/v/tests/use_alias_from_another_module_in_struct_field_test.v
+++ b/vlib/v/tests/use_alias_from_another_module_in_struct_field_test.v
@@ -1,0 +1,14 @@
+import alias_to_another_module
+
+struct MyStruct {
+	myfield alias_to_another_module.MyAlias
+}
+
+fn test_using_struct_with_alias() {
+	m := MyStruct{
+		myfield: alias_to_another_module.MyAlias{1, 2, 3}
+	}
+	dump(m)
+	assert m.myfield.alias_method() == 42
+	assert m.myfield.some_method() == 1005
+}

